packages/next/src/server/require-hook.js

```diff
@@ -32,21 +30,20 @@ mod._resolveFilename = function (
 // This is a hack to make sure that if a user requires a Next.js module that wasn't bundled
 // that needs to point to the rendering runtime version, it will point to the correct one.
 // This can happen on `pages` when a user requires a dependency that uses next/image for example.
-// This is only needed in production as in development we fallback to the external version.
-if (process.env.NODE_ENV !== 'development' && !process.env.TURBOPACK) {
-  mod.prototype.require = function (request) {
-    if (request.endsWith('.shared-runtime')) {
-      const isAppRequire = process.env.__NEXT_PRIVATE_RUNTIME_TYPE === 'app'
-      const currentRuntime = `${
-        isAppRequire
-          ? 'next/dist/compiled/next-server/app-page.runtime'
-          : 'next/dist/compiled/next-server/pages.runtime'
-      }.prod`
-      const base = path.basename(request, '.shared-runtime')
-      const camelized = base.replace(/-([a-z])/g, (g) => g[1].toUpperCase())
-      const instance = originalRequire.call(this, currentRuntime)
-      return instance.default.sharedModules[camelized]
-    }
-    return originalRequire.call(this, request)
+mod.prototype.require = function (request) {
+  if (
+    (process.env.NEXT_PHASE !== PHASE_PRODUCTION_BUILD ||
+      process.env.NEXT_IS_EXPORT_WORKER) &&
+    request.endsWith('.shared-runtime')
+  ) {
+    return originalRequire.call(
+      this,
+      `next/dist/server/future/route-modules/pages/vendored/contexts/${path.basename(
+        request,
+        '.shared-runtime'
+      )}`
+    )
```

Comment:
Should we consider having a dev only assertion that we never resolve react packages here for app router? In theory we always want any react* packages to be bundled so if we see a require for it here we know we've missed something?

Reply:
This call should only target files ending with `.shared-runtime` so it's unlikely react gets called here

Reply:
I meant should we hook into the cjs resolver and and warn if there is a require for `react-*`. The existence of such a require implies we missed something as part of our bundling. plus it would pick up the wrong package. It would be a different set of conditions inside the require override (or via _resolveFilename)